linode/terraform-provider-linode
# new: Generate temporary object storage keys to use

## 📝 Description

This change introduces an implicit object key generation in object storage bucket and object. If opt-in the temporary key generation, user don't have to provide their own obj access keys, and the keys will be created at apply-time with limited access to target bucket and cluster. The keys will be cleaned up automatically after use. User can only opt-in the temp key generation at provider-level.

This PR also combines #1337 which enables provider-level obj keys. With both change, now when object storage keys are required in the resources, we will look for them in the following order:
1) Whether the keys are specified in the resource configuration;
2) Whether the provider-level object keys exist;
3) Whether user opts-in temporary keys generation.

If none of these is satisfied, an error will be raised.

Addressed issue #1172.

## ✔️ How to Test

`make int-test PKG_NAME="linode/objbucket"`
`make int-test PKG_NAME="linode/obj"`

**Manual Test:**
1. In a sandbox environment, i.e. dx-devenv, run the following configuration to enable temporary keys generation when creating object storage bucket and an object in it
```terraform
provider "linode" {
    obj_use_temp_keys = true
}

resource "linode_object_storage_bucket" "test" {
    cluster = "us-mia-1"
    label = "test-obj-bucket"

    lifecycle_rule {
        prefix = "tf"
        enabled = true

        abort_incomplete_multipart_upload_days = 5

        expiration {
            date = "2024-03-21"
        }
    }
}

resource "linode_object_storage_object" "test" {
    bucket     = linode_object_storage_bucket.test.label
    cluster    = "us-mia-1"
    key        = "foobar"
    content = "cool"
}
```
2. Plan and apply the change. Observe the resources are created successfully.
